vendor/bat-native-ledger/src/bat/ledger/internal/endpoint/bitflyer/post_oauth/post_oauth_bitflyer.cc

```diff
@@ -0,0 +1,163 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/ledger/internal/endpoint/bitflyer/post_oauth/post_oauth_bitflyer.h"
+
+#include <utility>
+
+#include "base/base64url.h"
+#include "base/guid.h"
+#include "base/json/json_reader.h"
+#include "base/json/json_writer.h"
+#include "base/strings/stringprintf.h"
+#include "bat/ledger/internal/bitflyer/bitflyer_util.h"
+#include "bat/ledger/internal/endpoint/bitflyer/bitflyer_utils.h"
+#include "bat/ledger/internal/ledger_impl.h"
+#include "crypto/sha2.h"
+#include "net/http/http_status_code.h"
+
+using std::placeholders::_1;
+
+namespace ledger {
+namespace endpoint {
+namespace bitflyer {
+
+PostOauth::PostOauth(LedgerImpl* ledger) : ledger_(ledger) {
+  DCHECK(ledger_);
+}
+
+PostOauth::~PostOauth() = default;
+
+std::string PostOauth::GetUrl() {
+  return GetServerUrl("/api/link/v1/token");
+}
+
+std::string PostOauth::GeneratePayload(const std::string& external_account_id,
+                                       const std::string& code) {
+  const std::string client_id = ledger::bitflyer::GetClientId();
+  const std::string client_secret = GetClientSecret();
+  const std::string request_id = base::GenerateGUID();
+
+  base::DictionaryValue dict;
+  dict.SetStringKey("grant_type", "code");
+  dict.SetStringKey("code", code);
+  dict.SetStringKey("client_id", client_id);
+  dict.SetStringKey("client_secret", client_secret);
+  dict.SetIntKey("expires_in", 259002);
+  dict.SetStringKey("external_account_id", external_account_id);
+  dict.SetStringKey("request_id", request_id);
+  dict.SetStringKey("redirect_uri", "rewards://bitflyer/authorization");
+  dict.SetBoolKey("request_deposit_id", true);
+
+  // Always send PKCE code verifier and challenge
+  const std::string code_verifier =
+      ledger::bitflyer::GenerateRandomString(ledger::is_testing);
```

Comment:
https://tools.ietf.org/html/rfc7636#section-4.1 Specifies the `code verifier` should be between 43-128 characters in length (43-128 bytes). `GenerateRandomString` only generates 32 bytes of secure randomness which would be insufficient to match the specification.

Reply:
As a second note the `code verifier` should only contain the characters `a-z`, `A-Z`, `0-9`, `-`, `.`, `_`, `~` (ALPHA = %x41-5A / %x61-7A, DIGIT = %x30-39). `GenerateRandomString()` will return values between 0-255 which contains byte values outside that range.

Reply:
Good point about the length, will address that. Would base64-encoding the code verifier be an acceptable approach for your second point?

Reply:
Are you saying that we do the following? `crypto::RandBytes`->`base64`->`sha256`->`base64` (aka `base64` encode the randbytes?)

Reply:
Then feed them through the rest of the protocol?

Reply:
Yep, exactly.

Reply:
Should be fine. As long as the resulting `base64` encoding is between 43 and 128 long. So the `RandBytes` length will need to accommodate for that.

Reply:
OK, perfect.

Reply:
From the specification: 

> The client SHOULD create a "code_verifier" with a minimum of 256 bits
   of entropy.  This can be done by having a suitable random number
   generator create a 32-octet sequence.  The octet sequence can then be
   base64url-encoded to produce a 43-octet URL safe string to use as a
   "code_challenge" that has the required entropy.

So we can leave it as 32 and feed it into the base64url-encoder. 1 line change hopefully :)

Reply:
Works for me!

Reply:
This is fixed now.

Reply:
@orspetol Does this look good now from your point of view?